### ISSUE TITLE
Vetting our candidate list

### ISSUE THREAD

**Issue by gordonje** [california-civic-data-coalition/django-calaccess-processed-data#4]
We're currently deriving our list of candidates (combinations of persons, offices/district/agency sought and election year) from filers of the Form 501 (Candidate Intention Statement). So how comparable is our list to the what we see on the [CAL-ACCESS website](http://cal-access.ss.ca.gov/Campaign/Candidates/)?

The initial view of this page is the "certified" candidates for the 2016 general election, which lists:
- 40 State Senate candidates (2 candidates in each cell of a 2 X 10 grid)
- 159 Assembly Member candidates (80 races with as few as one and as many as three candidates. I basically just copy and pasted the contents of this table and did a little munging in Sublime Text).

Looking at the 2016 `election_year` records in our derived `calaccess_processed_candidate` table, we also have 20 distinct Senate races and 80 distinct Assembly races, but we have higher candidate counts in each race, presumably because were including the primary candidates.

So after switching the CAL-ACCESS candidates web page to prior elections, then 2016 primary, I see:
- 65 State Senate candidates, whereas we have 97.
- 228 Assembly Member candidates, whereas we have 294.

The State Senate district with the biggest discrepancy is 27. They have six candidates in that district and we have nine, the extras being:
- RICHARD M. MATHEWS (filer=id=[1387349](http://cal-access.ss.ca.gov/Campaign/Candidates/Detail.aspx?id=1387349), party=D)
  - Only has one `FILER_FILING_CD` record (`filing_id=2056473`), a Form 501 that has a 3/9/2015 `FILING_DATE` value. If you look up this `F501_502_CD` record, the `YR_OF_ELEC` value is 2016. Unfortunately, I get the CGI error whenever I try to [view the actual PDF](http://cal-access.ss.ca.gov/PDFGen/pdfgen.prg?filingid=2056473&amendid=0) of this filing.
  - You might think March 2015 is way too early to be filing to run in 2016, but there are plenty of other similar cases (18 in Mar 2015, 21 in Feb 2015, 34 in Jan 2015, 43 in Dec 2014...here's the [query and results](https://gist.github.com/gordonje/3975c652c02ae3b6fbe629e5c5ab0c44))
- GLORIA MERCADO-FORTINE (filer_id=[1376233](http://cal-access.ss.ca.gov/Campaign/Candidates/Detail.aspx?id=1376233), party=R)
  - Similar details to MATHEWS, but whereas his `F501_502_CD.COMMITTEE_ID` value is blank, her's is populated (1376232). This `FILER_ID` has eight `FILER_FILINGS_CD` records from 2015, and the last two filed in Nov have a `STMNT_TYPE` value of `10003` (i.e., `TERMINATION`).
- JIM DANTONA (filer_id=[1377460](http://cal-access.ss.ca.gov/Campaign/Candidates/Detail.aspx?id=1377460), party=D)
  - Similar MERCADO-FORTINE, his `F501_502_CD` record has a `COMMITTEE_ID` value (1377459), and the most recent `FILER_FILINGS_CD` records have a `TERMINATION` statement type.

So, tl;dr:
1. We should probably add an election type field to our derived candidates table. This would help in comparing our results to the CAL-ACCESS website.
2. Even after we do that, our results seem to include candidates without committees and with committees that terminated. But I'm not sure that's a bad thing.
3. We should probably also include the candidate's committeeID, and add a new model: `CandidateCommittee`.

**Comment by gordonje** [comment]
Also, @palewire was asking the question "Why are some of the candidate's names hyperlinked while some are not?"

For example, under the 2016 general election, State Senate 23,  O'DONNELL, RONALD J. is not hyperlinked. But you can actually still navigate to [his record](http://cal-access.ss.ca.gov/Campaign/Candidates/Detail.aspx?id=1363147) on the CAL-ACCESS website by passing `1363147` (his filerID) in to the `id` url parameter.

You'll see the message "This committee has not electronically filed a Form 460/461/450 for this election cycle."

Actually, there is a Form 460 `FILER_FILING_CD` record for his committee id (1368604) dated `2016-06-15`, but it has a `STMNT_STATUS` value of `11003` (i.e., `'NEEDS REVIEW'`).

**Comment by gordonje** [comment]
Ugh. Adding the type of election to the result set is not as straight forward as I thought because `F501_502_CD.ELEC_TYPE` is so sparsely populated. Apparently there have only been 64 general election candidates in the entire history of the system.

Ugh.

**Comment by palewire** [comment]
Huh. They seem to record election results on the public website, so I wonder if this is another one of those things they don't include in the dump.

**Comment by gordonje** [comment]
So actually, reading the fine print of the forms more carefully, I don't believe every candidate will have a `F501_502_CD` record, which would help explain the discrepancy between these two numbers:

* **9,822**: Distinct filers of Form 501 (Candidate Intention Statements). Presumably there should be only one of these for each combination of person and office sought in a given election year (after you sort out the amended filings, of course):

```sql
SELECT COUNT(DISTINCT "FILER_ID")
FROM "F501_502_CD"
-- exclude Form 502 (Campaign Bank Account Statement) filings 
WHERE "FORM_TYPE" = 'F501'; 
```

* **18,161**: Distinct filers with the filer_type `'CANDIDATE/OFFICEHOLDER'`:

```sql
select COUNT(DISTINCT "FILER_ID") 
from "FILER_TO_FILER_TYPE_CD"
-- DESCRIPTION of FILER_TYPE.filer_type = 8 is 'CANDIDATE/OFFICEHOLDER'
where "FILER_TYPE" = 8; 
```

From the [Form 501](http://calaccess.californiacivicdata.org/forms/f501/) (Candidate Intention Statement):
> 
A candidate for state or local office must file this form for each election, including reelection to the same office. Exception: Candidates for county central committee that do not raise or spend $2,000 or more in a calendar year.

From the [Form 470](http://calaccess.californiacivicdata.org/forms/f470/) (Officeholder and Candidate Campaign Statement, Short Form):
>
Form 470 is for use by officeholders and candidates who:
• do not have a controlled committee;
• do not anticipate receiving contributions totaling $2,000 or more during the calendar year; and
• do not anticipate spending $2,000 or more during the calendar year.
Officeholders and candidates who have a controlled committee or who have raised or spent $2,000, file the Recipient Committee Statement – Form 460.
Exceptions:
The following individuals seeking or holding office are not required to ﬁle campaign disclosure statements (Form 470 or Form 460):
• candidates for county central committee offices that do not raise or spend $2,000 or more in a
calendar year;
• officeholders whose salaries are less than $200 per month and judicial candidates who have not made or received contributions or made expenditures during non-election years; and
• judges who do not receive contributions and who make personal expenditures of less than $1,000 or more in non-election years.

From the [Form 460](http://calaccess.californiacivicdata.org/forms/f460/) (Recipient Committee Campaign Statement):
>The Form 460 is for use by all recipient committees, including:
Candidates, Officeholders and Their Controlled Committees
•  A candidate or officeholder who has a controlled committee, or who has raised or spent or will raise or spend $2,000 or more during a calendar year in connection with election to office or holding office.  The Form 460 is also required if $2,000 or more will be raised or spent during the calendar year at the behest of the officeholder or candidate.

...
>
Primarily Formed Candidate/Officeholder Committees
•  A person, entity, or organization that receives contributions totaling $2,000 or more during a calendar year to support or oppose a single candidate or officeholder, or two or more candidates or officeholders who are being voted upon in the same city, county, or multi-county election.  This type of committee is not controlled by the candidate(s) or officeholder(s).

Use the forms...

**Comment by gordonje** [comment]
There are 22 CAL-ACCESS raw-data tables (in the downloaded ZIP) that include a `FILER_ID` field. If you [stack all these columns](https://gist.github.com/gordonje/414dd0b7b9b3dd13d2fa13986b2c3725) on top of each other, there are 364,195 unique `FILER_ID` values in the database.

Yet [not one of these 22 tables](https://gist.github.com/gordonje/61d438ace267921d8cee04ad30bc099d) includes all unique `FILER_ID`s. One would expect to find the all in `FILERS_CD`, but that table is 12,153 shy.

Here's the breakdown of [mystery filer_id values by table](https://gist.github.com/gordonje/6c6afd6bf062ac781a802c4cc7c171e7) is here:

```
table_name                  data_type          filer_id_count  
--------------------------  -----------------  --------------  
EFS_FILING_LOG_CD           character varying  4818            
F501_502_CD                 character varying  4494            
CVR_LOBBY_DISCLOSURE_CD     character varying  3614            
CVR_CAMPAIGN_DISCLOSURE_CD  character varying  2703            
CVR_REGISTRATION_CD         character varying  2693            
CVR_SO_CD                   character varying  1000            
LOBBYING_CHG_LOG_CD         integer            199             
RECEIVED_FILINGS_CD         integer            132             
LOBBYIST_CONTRIBUTIONS3_CD  integer            2               
LOBBYIST_CONTRIBUTIONS2_CD  integer            2               
LOBBYIST_CONTRIBUTIONS1_CD  integer            2    
```

Probably no coincidence that this list includes all non-empty tables with a `FILER_ID` column that allows non-numeric characters.

We don't quite have the same problem for the `FILING_ID`s. There are 38 released raw-data tables with a `FILING_ID` column (all with the same data type, integer AWESOME). If you [stack all of these columns](https://gist.github.com/gordonje/5bfe3773f9ba087e65a67bd0f24985d2), there are 1,770,380 unique `FILING_ID` values (broken down by table [here](https://gist.github.com/gordonje/8e9a93bd9ea924c72446309a943dbc56)), and [all but 218](https://gist.github.com/gordonje/6d0b02c611f14898018d9f7456b779bf) are in the `FILINGS_CD` table. Those missing `FILING_ID`s are in `RECEIVED_FILINGS_CD`, most of which have `"Filing accepted"` in the comments (weird that they would not then be in `FILINGS_CD`).

tl;dr:

Seems like we will need to build the list of filers, including the candidates, up from the various filing tables.

**Comment by gordonje** [comment]
What I'm still unclear about is whether or not is if values in `FILER_XREF_CD.XREF_ID` can show up as values in the `FILER_ID` columns on the various filings tables.

The MapLight processing sql seem to suggest that this is true. Their cleaning process includes [creating a translator](https://github.com/maplight/CAPS/blob/master/scripts/process_stage_1.sql#L151-L163) of what they call `disclosure_filer_id` `filer_id`. The `disclosure_filer_id` is the value from `CVR_CAMPAIGN_DISCLOSURE.FILER_ID` and the `filer_id` is either `FILER_XREF_CD.FILER_ID` or `FILER_XREF_CD.XREF_ID`.

But I think there's a problem with how they are going about this: They are [assigning an integer data type](https://github.com/maplight/CAPS/blob/master/scripts/install_ftp_tables.sql#L808) to `FILER_XREF_CD.XREF_ID` when they create their version of the `FILER_XREF_CD` table, but it contains [over 13k values](https://gist.github.com/gordonje/35f6e6b26d0f57f899c788da31aac8bf) with non-numeric characters.

**Comment by palewire** [comment]
We have tests for this now.
